Add alternative implementation for Mojang's FileIO.

cybertiger opened the pull request:
This probably needs more extensive testing before pulling.

New IO implementation is entirely optional.
Better locking.
Better guarantees about closing region files on world unload.
Better detection and handling of corrupt region files.
Region files can shrink as well as grow.
Better free space map implementation for region files.
Fairly minimal diff.
Configurable thread count for async chunk saving.
Configurable limit to the async chunk saving queue size.

Future: possibility of a plugin API to modify world persistence.
Future: possibility of per world disk IO stats.

cybertiger committed:
Add alternative implementation for Mojang's FileIO.

New IO implementation is entirely optional.
Better locking.
Better guarantees about closing region files on world unload.
Better detection and handling of corrupt region files.
Region files can shrink as well as grow.
Better free space map implementation for region files.
Fairly minimal diff.
Configurable thread count for async chunk saving.
Configurable limit to the async chunk saving queue size.

Future: possibility of a plugin API to modify world persistence.
Future: possibility of per world disk IO stats.

cybertiger committed:
Fix a couple of minor TODO relating to player saves.

Reimplement NMS renaming logic with saving player files (may prevent data loss)
Reimplement NMS logic to use offline player files when an online player file cannot be found (and server is in online mode).

cybertiger committed:
Add disk io stats via the iostat command.

kashike commented on Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch:
```diff
@@ -0,0 +1,739 @@
++            return true;
++        }
++
++        if ( !PaperConfig.usePaperIO ) {
```
No need for weird formatting.

aikar replied:
Also IOStatCommand for consistency.

kashike commented on Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch:
```diff
@@ -0,0 +1,739 @@
++
++    @Override
++    public boolean execute(CommandSender sender, String commandLabel, String[] args) {
++        if ( !testPermission(sender) ) {
```
No need for weird formatting.

kashike commented on Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch:
```diff
@@ -0,0 +1,739 @@
++            return true;
++        }
++
++        if ( !PaperConfig.paperIOStats ) {
```
No need for weird formatting.

kashike commented on Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch:
```diff
@@ -0,0 +1,739 @@
+ 
+-    // TODO
+-    // Counters pertaining to world saving, so we can get per world stats on chunk io
++    public void bytesRead(long bytes);
```
`public` is not needed.

kashike commented on Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch:
```diff
@@ -0,0 +1,739 @@
++ * To change this license header, choose License Headers in Project Properties.
++ * To change this template file, choose Tools | Templates
++ * and open the template in the editor.
++ */
```
????????

kashike commented on Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch:
```diff
@@ -0,0 +1,739 @@
++ *
++ * @author antony
++ */
++public class NullPersistenceCounters implements PersistenceCounters {
```
`final`?

cybertiger committed:
Fix a bug where plugin loaded worlds didn't use the alternate IO.

cybertiger committed:
Add async chunk save queue size to iostat command.

Includes fixed formatting issues as pointed out on the PR request.

cybertiger committed:
Add 1,5,15 minute running averages to iostat for chunks and bytes.

cybertiger committed:
Fix issue with 15 minute average always been zero after more than 15 minutes.

cybertiger committed:
Multiple changes.

Move paper-io settings to paper-io section of config.
Add setting to disable automatic fixing of corrupt chunks.
Refactor AnvilRegionFile to use classic IO for everything (no memory mapped io)
Fixed a bug with allocating new region files not writing a clean header.
Fixed a stupid bug with checking if a previously saved chunk is present.
Remove boolean flag on ChunkException for 'corrupt', instead just return null if
 someone tries to load a chunk which does not exist.
